src/components/tl/ucp/tl_ucp_team.c

```diff
@@ -0,0 +1,243 @@
+        }
+        storage->state = UCC_TL_UCP_ADDR_EXCHANGE_GATHER;
+        return UCC_INPROGRESS;
+    case UCC_TL_UCP_ADDR_EXCHANGE_GATHER:
+        storage->state = UCC_TL_UCP_ADDR_EXCHANGE_COMPLETE;
+        break;
+    }
+    return UCC_OK;
+
+err:
+    free(storage->addrlens);
+    free(storage->addresses);
+    free(storage);
+    return status;
+}
+
+void ucc_tl_ucp_addr_storage_free(ucc_tl_ucp_addr_storage_t *storage)
+{
+    free(storage->addresses);
+    ucc_assert(NULL == storage->addrlens);
+    free(storage);
+}
+
+UCC_CLASS_INIT_FUNC(ucc_tl_ucp_team_t, ucc_base_context_t *tl_context,
+                    const ucc_base_team_params_t *params)
+{
+    ucc_status_t          status = UCC_OK;
+    ucc_tl_ucp_context_t *ctx =
+        ucc_derived_of(tl_context, ucc_tl_ucp_context_t);
+    UCC_CLASS_CALL_SUPER_INIT(ucc_tl_team_t, &ctx->super);
+    /* TODO: init based on ctx settings and on params: need to check
+             if all the necessary ranks mappings are provided */
+    self->context_ep_storage = 0;
+    self->addr_storage       = NULL;
+    self->size               = params->params.oob.participants;
+    if (self->context_ep_storage) {
+        self->status = UCC_OK;
+    } else {
+        self->status = UCC_INPROGRESS;
+        status       = ucc_tl_ucp_addr_exchange_start(ctx, params->params.oob,
+                                                &self->addr_storage);
+    }
+    tl_info(tl_context->lib, "posted tl team: %p", self);
+    return status;
+}
+
+UCC_CLASS_CLEANUP_FUNC(ucc_tl_ucp_team_t)
+{
+    if (self->addr_storage) {
+        ucc_tl_ucp_addr_storage_free(self->addr_storage);
+    }
+    tl_info(self->super.super.context->lib, "finalizing tl team: %p", self);
+}
+
+UCC_CLASS_DEFINE_DELETE_FUNC(ucc_tl_ucp_team_t, ucc_base_team_t);
+UCC_CLASS_DEFINE(ucc_tl_ucp_team_t, ucc_tl_team_t);
+
+ucc_status_t ucc_tl_ucp_team_destroy(ucc_base_team_t *tl_team)
+{
+    ucc_tl_ucp_team_t    *team = ucc_derived_of(tl_team, ucc_tl_ucp_team_t);
+    ucc_tl_ucp_context_t *ctx  = UCC_TL_UCP_TEAM_CTX(team);
+    ucc_status_t          status;
+    if (team->eps) {
+        status = ucc_tl_ucp_close_eps(ctx, team->eps, team->size);
+        if (UCC_INPROGRESS == status) {
+            return status;
+        } else if (UCC_OK != status) {
+            tl_error(team->super.super.context->lib,
+                     "failed to close team eps");
+            return status;
+        }
+    }
+    UCC_CLASS_DELETE_FUNC_NAME(ucc_tl_ucp_team_t)(tl_team);
+    return UCC_OK;
+}
+
+static ucc_status_t ucc_tl_ucp_team_preconnect(ucc_tl_ucp_team_t *team)
+{
+    ucc_tl_ucp_context_t *ctx = UCC_TL_UCP_TEAM_CTX(team);
+    int                   i;
+    ucc_status_t          status;
+    for (i = 0; i < team->size; i++) {
+        status = ucc_tl_ucp_connect_ep(ctx, team, team->addr_storage->addresses,
+                                       team->addr_storage->max_addrlen, i);
+        if (UCC_OK != status) {
+            return status;
```

Comment:
Do we need to close all already opened endpoints in case of failure?

Reply:
i think we should. 👍